design.md

```diff
@@ -0,0 +1,128 @@
+
+#### GitFS
+
+- Each repo is uniquely identified by a Git Repo URI.
+- Each file in a repo is uniquely identified by a Git File URI.
+- Each Git URI includes a remote.
+- Operations:
+  - Clone or pull a repo.
+  - List files in a repo.
+  - Read the contents of a file in a repo.
+  - Run a script in a repo.
+    - May pass arguments as a list of objects.
+    - May receive a return value as an object.
+  - Load a robot from a robot config file in a repo.
+
+#### Hardware Management
+
+- Operations:
+  - Load a robot from its config file.
+    - May specify the contents of the config file.
+    - May specify a Git File URI.
+    - Must specify whether to connect to hardware or a simulator.
+      - In the hardware case, a connection method must be specified.
+      - In the simulator case, a simulator plugin must be specified.
+
+#### Simulator Plugins
+
+- Operations:
+  - List available plugins
+  - Add a plugin
+    - Adding a plugin with an ID that conflicts with an ID of an already-added plugin throws an error.
+    - ! RFC: What should the schema for adding a plugin be? The most natural method is by specifying the path to the Jar file, but we don't want to use file paths. We could support specifying a Git File URI. What else? I don't want to allow specifying content because that would involve sending potentially lots of megabytes over the RPC call.
+  - Remove a plugin by its ID.
+- Each plugin has a unique ID and a Jar file.
+- The kernel must be able to download simulator plugins.
+  - We can ship the default simulator plugin with the IntelliJ plugin, but there must be a way to download any of them because of the [usability requirement](#usability).
+
+#### UI Interaction
+
+- ! RFC: The kernel must be able to notify the UI of multiple tasks happening in parallel. For example, running a script could cause some repos to get cloned and a toolchain to get downloaded in parallel. As a user, I want to see both of these operations as tasks in IntelliJ happening in parallel. As a general rule, a script can start multiple tasks.
+- ! RFC: The kernel needs a way to ask the UI for confirmation with a custom confirmation message. For example, if the kernel needs to flash the device with a new program, by default it should ask the user for confirmation. There also needs to be some local environment configuration option that bypasses this so that the confirmation results is either always allowed or denied (for headless operation). For example, you may want to always allow the kernel to flash the device during development or during some competitions (as a user, I would rather have a delayed start to my match rather than having the robot not work at all; it depends on the competition). As another example, you may want to always stop the kernel from flashing the device if you have released a product and you expect all the products to work the same for all your users and would rather handle these issues with your own support system.
```

Comment:
I would like people more familiar with gRPC to provide some input here on how we can design this.

Reply:
Assumption: any of these items (on either point) only happen in the context of "running a script". If not, this still works, but would require some tweaking.

I'd recommend making the "run script" call a [bidirectional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition). This essentially means that the call (initiated by the client) opens two independent, strictly ordered streams of objects (one in either direction). The object type of each stream can be different, and those object types can effectively be sum types by use of the `oneOf` construct in protobuf. The resulting communication might look like this (C = client, K = kernel):

C: *Initiate call*
K: *Accept call (done by gRPC system)*
(Streams are now open)
C: Run script: path "git:a/b/c", request ID 0 *(to match the resulting task created by the kernel to this request)*
K: Task started for request 0: task ID 0 *(to match future updates to this info)*, description "Lorum ipsum", progress NaN *(no way to know progress of a script; this would show an indeterminate progress bar in IntelliJ, for example)*
(Script requires a repo and a toolchain to be downloaded)
K: Task created by request of task 0: task ID 1, description "Download github/...", progress 0.0
K: Task created by request of task 0: task ID 2, description "Download toolchain arm-none-eabi", progress 0.0
K: Task update, task ID 1, progress 0.31
K: Task update, task ID 2, progress 0.25
(etc.)
K: Task complete, task ID 1
K: Task update, task ID 2, progress 0.89
K: Task complete, task ID 2
(Script calls API to set its description - if we want to support that)
K: Task update, task ID 0, description "Example"
(Script attempts to flash device, no override in environment)
K: User query, question "Please confirm flashing device XYZ", type "OkCancel", query ID 0
(User clicks OK)
C: User query response, query ID 0, answer "OK"
(Do the task thing again for flashing, then rest of script runs, etc.)
K: Task complete, Task ID 0
K: *Close stream*
C: *Close stream*
(Call is now over)